### PR TITLE
chore: canonicalize constants

### PR DESCRIPTION
**Pull request by tobiasgrosser**

### CODE REVIEW

**Comment by tobiasgrosser** [comment]
I pulled this PR out of https://github.com/opencompl/ssa/pull/356, as this seems to be a self-contained change that can be verified rather mechanically. In https://github.com/opencompl/ssa/pull/356, I made some of these statements new simp-lemma on BitVec which caused unrelated changes in BitVec.

**Comment by tobiasgrosser** [comment]
> Alive Statistics: 30 / 93 (63 failed)
Nice. These still need to be fixed.

**Comment by tobiasgrosser** [comment]
Nice. This is now ready to review (and merge). It should reduce a lot of the noise we saw in https://github.com/opencompl/ssa/pull/356.

File: SSA/Projects/InstCombine/ForStd.lean [hunk 1]
```diff
@@ -131,4 +131,11 @@ theorem bitvec_minus_one : BitVec.ofInt w (Int.negSucc 0) = (-1 : _root_.BitVec
     rw [zeroBitwidth]
     simp
 
+theorem bitvec_minus_one' : BitVec.ofInt w (-1) = (-1 : BitVec w) := by
+  have x : -1 = Int.negSucc 0 := by simp
+  rw [x, bitvec_minus_one]
+
+
+
+
```
Comment (bollu) [inline SSA/Projects/InstCombine/ForStd.lean:140]:
```suggestion
theorem bitvec_minus_one' : BitVec.ofInt w (-1) = (-1 : BitVec w) := by
  have x : -1 = Int.negSucc 0 := by simp
  rw [x, bitvec_minus_one]
  
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 2]
```diff
@@ -299,22 +298,18 @@ def sdiv_one_one' (h : 1 < w) : BitVec.sdiv 1#w 1#w = 1#w := by
   apply BitVec.eq_of_toNat_eq
   simp [hone]
 
-def sdiv_one_one : BitVec.sdiv (BitVec.ofInt w 1) 1 = 1 := by
+def sdiv_one_one : BitVec.sdiv 1#w 1 = 1 := by
```
Comment (bollu) [inline SSA/Projects/InstCombine/ForLean.lean:301]: This is not in simp-normal-form for sure, right? It's depending on the `1 : Nat` to get coerced into `1#w : BitVec w`? A good way to check that the theorem statement is in simp normal form is to write 

```suggestion
def sdiv_one_one : BitVec.sdiv 1#w 1 = 1 := by simp
```

and see if `simp` complains about the theorem statement being in normal form. I would imagine that the canonical theorem statement is

```lean
def sdiv_one_one : BitVec.sdiv 1#w 1#w = 1#w := by simp
```

File: SSA/Projects/InstCombine/ForLean.lean [hunk 3]
```diff
@@ -123,9 +123,10 @@ lemma ofInt_ofNat (w n : Nat) :
     BitVec.ofInt w (OfNat.ofNat n) = BitVec.ofNat w n :=
   rfl
 
+lemma ofInt_ofNat' : BitVec.ofInt w (OfNat.ofNat (α := ℤ) x ) = x#w := rfl
```
Comment (bollu) [inline SSA/Projects/InstCombine/ForLean.lean:126]: Giving types here should just work:

```suggestion
lemma ofInt_ofNat' {x : Nat} : BitVec.ofInt w (OfNat.ofNat x) = x#w := rfl
```

I am once again suspicious of this, though. Doesn't the `OfNat` simp away?

**Review by bollu — CHANGES_REQUESTED** [review]
I think the theorem statements need to be in `simp` normal form before we can merge this. If they are not being normalized correctly, then we ought to write the correct simps to go between `Int/Nat/OfNat.ofNat`

File: SSA/Projects/InstCombine/ForStd.lean (hunk 1 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForStd.lean:140]: @bollu, what is the change you are suggesting here?

Comment (bollu) [reply SSA/Projects/InstCombine/ForStd.lean:140]: @tobiasgrosser aha,  I wanted to nuke the extra newlines, but it seems like github's "suggestion" algorithm strips trailing newline, so it looks like I'm suggesting we add a newline..

File: SSA/Projects/InstCombine/ForLean.lean (hunk 2 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean]: Resolved.

File: SSA/Projects/InstCombine/ForStd.lean (hunk 1 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForStd.lean]: Done.

File: SSA/Projects/InstCombine/ForLean.lean (hunk 3 above)
Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean:126]: We already have `ofInt_ofNat`, but it does not apply. The problem is here:

```
+++ b/SSA/Projects/InstCombine/AliveHandwrittenLargeExamples.lean
@@ -232,6 +232,8 @@ def alive_simplifyMulDivRem805' (w : Nat) :
   simp
   split_ifs with c
   simp
+  -- TODO: This one does not work: simp [(BitVec.ofInt_ofNat)]
+  simp [(BitVec.ofInt_ofNat')]
```

Comment (alexkeizer) [reply SSA/Projects/InstCombine/ForLean.lean:126]: I'm very confused, `n#w` is a macro that expands to `BitVec.ofNat`, so `ofInt_ofNat` and `ofInt_ofNat'` should be the same theorem (as in, syntactically equal statements, post elaboration)

Comment (bollu) [reply SSA/Projects/InstCombine/ForLean.lean:126]: The upstream issue tracking this is https://github.com/leanprover/lean4/issues/2867

**Comment by bollu** [comment]
Lgtm

**Review by bollu — APPROVED** [review]

File: SSA/Projects/InstCombine/ForLean.lean [hunk 4]
```diff
@@ -120,9 +120,11 @@ theorem Nat.one_mod_two_pow_succ_eq {n : Nat} : 1 % 2 ^ n.succ = 1 := by
 
 @[simp]
 lemma ofInt_ofNat (w n : Nat) :
-    BitVec.ofInt w (OfNat.ofNat n) = BitVec.ofNat w n :=
+    BitVec.ofInt w (no_index (OfNat.ofNat n)) = BitVec.ofNat w n :=
   rfl
 
+lemma ofInt_ofNat' : BitVec.ofInt w (OfNat.ofNat (α := ℤ) x ) = x#w := rfl
+
 -- @[simp]
```
Comment (alexkeizer) [inline SSA/Projects/InstCombine/ForLean.lean:128]: It seems you didn't remove the primed version, but this one should not be needed after the `no_index` fix, right?

Comment (tobiasgrosser) [reply SSA/Projects/InstCombine/ForLean.lean:128]: Right. Sorry for this and thank you for the post-commit review: https://github.com/opencompl/ssa/pull/new/clean_unneeded_stmt.

**Pull request merged**